src/fundus/scraping/common_crawl/pipeline.py

```diff
@@ -0,0 +1,274 @@
+
+        if start >= end:
+            raise ValueError("Start date has to be < end date.")
+
+        if start < datetime(2016, 8, 1):
+            raise ValueError("The default, and earliest possible, start date is 2016/08/01.")
+
+        if end > datetime.now():
+            raise ValueError("The specified end date is in the future. We don't want to give spoilers, do we?")
+
+        date_sequence: List[datetime] = [dt for dt in rrule(MONTHLY, dtstart=start, until=end)]
+        urls = [
+            self.server_address + f"crawl-data/CC-NEWS/{date.strftime('%Y/%m')}/warc.paths.gz" for date in date_sequence
+        ]
+
+        def load_paths(url: str) -> List[str]:
+            with requests.Session() as session:
+                paths = gzip.decompress(session.get(url).content).decode("utf-8").split()
+                bar.update()
+                return paths
+
+        with ThreadPool(processes=len(urls)) as pool, tqdm(total=len(urls), desc="Load WARC paths", leave=False) as bar:
+            warc_paths = more_itertools.flatten(pool.map(load_paths, urls))
+
+        start_strf = start.strftime("%Y%m%d%H%M%S")
+        end_strf = end.strftime("%Y%m%d%H%M%S")
+
+        def filter_warc_path_by_date(path: str) -> bool:
+            if match := date_pattern.search(path):
+                return start_strf <= match["date"] <= end_strf
+            else:
+                return False
+
+        return sorted(
+            [self.server_address + warc_path for warc_path in filter(filter_warc_path_by_date, warc_paths)],
+            reverse=True,
+        )
+
+    @staticmethod
+    def _fetch_articles(
+        warc_path: str,
+        publishers: Tuple[PublisherEnum],
+        error_handling: Literal["suppress", "catch", "raise"],
+        extraction_filter: Optional[ExtractionFilter] = None,
+        url_filter: Optional[URLFilter] = None,
+    ) -> Iterator[Article]:
+        source = CCNewsSource(*publishers, warc_path=warc_path)
+        scraper = CCNewsScraper(source)
+        for article in scraper.scrape(error_handling, extraction_filter, url_filter):
+            yield article
+
+    @staticmethod
+    def _queue_wrapper(
+        queue: Queue[_T],
+        target: Callable[P, Iterator[_T]],
+    ) -> Callable[P, None]:
+        def wrapper(*args, **kwargs) -> None:
+            for obj in target(*args, **kwargs):
+                queue.put(obj)
+
+        return wrapper
+
+    @staticmethod
+    def _single_crawl(warc_paths: List[str], target: Callable[[str], Iterator[Article]]) -> Iterator[Article]:
+        for warc_path in warc_paths:
+            yield from target(warc_path)
+
+    def _parallel_crawl(self, warc_paths: List[str], target: Callable[[str], Iterator[Article]]) -> Iterator[Article]:
+        with Manager() as manager, Pool(processes=min(self.processes, len(warc_paths))) as pool:
+            article_queue: Queue[Article] = manager.Queue()
```

Comment:
As you pointed out it's not possible to use `multiprocessing.Queue` here since it is unpickable. The sole function of  `multiprocessing.Manager` is to provide pickable proxies for objects to "share" objects between processes. What happens within the manager is, that not the actual object is shared, but the object's attributes (methods, properties, etc) are exposed through the proxy. So when calling a method of the proxy, the manager passes the call to the managed object instance.

Reply:
Ah, that's also why we access the queue with `manager.Queue()`. Thanks for the insights :).